cloudvolume/datasource/precomputed/sharding.py

```diff
@@ -740,10 +739,11 @@ def synthesize_shard_file(spec, label_group, progress=False, presorted=False):
       minishard_index[0, i] = label - last_label
       minishard_index[1, i] = 0 # minishard_index[2, i - 1]
       minishard_index[2, i] = len(binary)
-      minishard += binary
+      minishard_components.append(binary)
       last_label = label
       del minishardgrp[label]
-    
+
+    minishard = b"".join(minishard_components)
```

Comment:
This is cool, but is this validated by performance tests?

Reply:
Yeah, that one is/was the biggest bottleneck:

With `x = b'a' * 1000`
```python
res = b''
for i in range(1000):
  res += x
```
`34.6 ms ± 505 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)`  - which makes sense; strings are immutable, so you end up allocating/discarding a new copy every time.

```python
tmp = []
for i in range(1000):
  tmp.append(x)
res = b''.join(tmp)
```
`147 µs ± 1.34 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)`

Pretty sure it's related to https://stackoverflow.com/questions/1349311/python-string-join-is-faster-than-but-whats-wrong-here/1350289#1350289
And I can completely understand his rant, because if you do that same experiment above with `str` instead of `bytes`, it appears to take that magical optimization path he is talking about and you suddenly end up with similar performance for both methods (with `join` still slightly faster, though)

Using list comprehension would be slightly faster again, but with the change above, most of the time is now spent copying and compressing the minishard components

Reply:
How long in wall clock time was make shard taking you then and now? Thanks for the info, didn't know that about strings. I think I take the side of the optimizers as real newbies aren't going to be researching the right way to do it. In this case, I guess I'm not a newbie, but I just figured I'd live with whatever performance += would give me until it became a crisis. Thanks for looking into it!

Reply:
Don't have the exact numbers at hand. Can check once the cluster is free again. `make_shard` went from approximately 2000 s to 500 s for my 8192x8192x512 vx shards (raw+gzip). Most of that is spent on gzipping each of the 32,768 tiny 256x256x16 components.

Another ~60 s is numpy [copying those cutouts into contiguous arrays](https://github.com/seung-lab/cloud-volume/blob/master/cloudvolume/chunks.py#L137)

Reply:
That's a big volume.... those numbers sound reasonable.

Reply:
I think certain brotli compression levels are faster than gzip for approximately the same compression. CloudFiles now supports brotli for all supported providers.

Reply:
That's correct, although currently not supported in Neuroglancer - for sharded volumes, decoding is done by the [async_computation workers](https://github.com/seung-lab/neuroglancer/tree/master/src/neuroglancer/async_computation).